File: leetcode/problems/language-specific/go/97.interleaving-string.go
/*
 * @lc app=leetcode id=97 lang=golang
 *
 * [97] Interleaving String
 */

// @lc code=start

// * Bottom-Up Iterative Solution | O(mn) Time | O(mn) Space
// * m -> Length of s1 | n -> Length of s2

func isInterleave(s1 string, s2 string, s3 string) bool {
	s1Size, s2Size, s3Size := len(s1), len(s2), len(s3)
	if s1Size+s2Size != s3Size {
		return false
	}

	dp := make([][]bool, s1Size+1)
	for i := 0; i < len(dp); i++ {
		dp[i] = make([]bool, s2Size+1)
	}

	dp[s1Size][s2Size] = true
	for i := s1Size; i >= 0; i-- {
		for j := s2Size; j >= 0; j-- {
			if i < s1Size && s1[i] == s3[i+j] && dp[i+1][j] {
				dp[i][j] = true
			}
			if j < s2Size && s2[j] == s3[i+j] && dp[i][j+1] {
				dp[i][j] = true
			}
		}
	}

	return dp[0][0]
}

// @lc code=end

File: leetcode/problems/sessions/2023-05-01_2023-??-??/97.interleaving-string.py
#
# @lc app=leetcode id=97 lang=python3
#
# [97] Interleaving String
#

# @lc code=start

# * Iterative Bottom-Up Solution | O(mn) Time | O(mn) Space
# * m -> The length of s1 string | n -> The length of s2 string


class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        s1_len, s2_len, s3_len = len(s1), len(s2), len(s3)
        if s1_len + s2_len != s3_len:
            return False

        dp = [[False] * (s2_len + 1) for _ in range(s1_len + 1)]
        dp[s1_len][s2_len] = True
        for i in range(s1_len, -1, -1):
            for j in range(s2_len, -1, -1):
                if i < s1_len and s1[i] == s3[i + j] and dp[i + 1][j]:
                    dp[i][j] = True
                if j < s2_len and s2[j] == s3[i + j] and dp[i][j + 1]:
                    dp[i][j] = True

        return dp[0][0]


# @lc code=end
